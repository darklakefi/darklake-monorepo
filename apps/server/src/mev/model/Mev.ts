import { IsEnum, IsIn, IsInt, IsOptional, Max, Min, Validate } from "class-validator";
import { IsValidSolanaAddress } from "../../validator/IsValidSolanaAddress";
import { Type } from "class-transformer";
import { SandwichEvent } from "@prisma/client";
import { formatSolAmount } from "src/utils/blockchain";

export class GetMevTotalExtractedQuery {
  @Validate(IsValidSolanaAddress)
  address: string;
}

export interface MevTotalExtracted {
  totalSolExtracted: number;
  totalUsdExtracted?: number;
}

export interface GetMevTotalExtractedResponse {
  data?: MevTotalExtracted;
  processingBlocks: {
    completed: number;
    total: number;
  };
}

export enum MevAttacksOrderBy {
  "DATE" = "date",
  "AMOUNT_DRAINED" = "amountDrained",
}

export class GetMevAttacksQuery {
  @Validate(IsValidSolanaAddress)
  address: string;

  @IsOptional()
  @IsInt()
  @Min(1)
  @Max(100)
  @Type(() => Number)
  limit?: number;

  @IsOptional()
  @IsInt()
  @Type(() => Number)
  offset?: number;

  @IsOptional()
  @IsEnum(MevAttacksOrderBy)
  orderBy?: MevAttacksOrderBy;

  @IsOptional()
  @IsIn(["asc", "desc"])
  direction?: "asc" | "desc";
}

export interface Mev {
  tokenName: string;
  timestamp: number;
  solAmountSent: number;
  solAmountDrained: number;
  transactionHash: string;
}

export enum MevAttackSwapType {
  BUY = "BUY",
  SELL = "SELL",
}

export interface MevTransaction {
  address: string;
  signature: string;
}

export class MevAttackResponse {
  swapType: MevAttackSwapType;
  tokenName: string;
  timestamp: number;
  solAmount: {
    sent: number;
    lost: number;
  };
  transactions: {
    frontRun: MevTransaction;
    victim: MevTransaction;
    backRun: MevTransaction;
  };

  constructor(props: SandwichEvent) {
    this.solAmount = {
      sent: formatSolAmount(props.sol_amount_swap),
      lost: formatSolAmount(props.sol_amount_drained),
    };

    this.tokenName = "UNKNOWN"; // TODO: add token metadata service & get token name
    this.timestamp = +props.occurred_at;
    this.transactions = {
      frontRun: {
        address: "0x0", // TODO: add address
        signature: "0x0",
      },
      victim: {
        address: "0x0",
        signature: props.tx_hash_victim_swap,
      },
      backRun: {
        address: "0x0",
        signature: "0x0",
      },
    };
  }
}
