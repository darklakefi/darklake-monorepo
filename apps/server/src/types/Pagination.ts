export type PaginatedResponse<T> = {
  limit: number;
  offset: number;
  total: number;
  result: T;
};

export enum PaginatedResponseDataLimit {
  MEV_EVENTS_LIST = 100,
}
