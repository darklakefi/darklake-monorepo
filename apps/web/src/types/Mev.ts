export enum MevAttackSwapType {
  BUY = "BUY",
  SELL = "SELL",
}

export interface MevTransaction {
  address: string;
  signature: string;
}

export interface MevAttack {
  swapType: MevAttackSwapType;
  tokenName: string;
  timestamp: number;
  solAmount: {
    sent: number;
    lost: number;
  };
  transactions: {
    frontRun: MevTransaction;
    victim: MevTransaction;
    backRun: MevTransaction;
  };
}

export interface MevTotalExtracted {
  totalSolExtracted: number;
  totalUsdExtracted?: number;
}

export interface GetMevTotalExtractedResponse {
  data?: MevTotalExtracted;
  processingBlocks: {
    completed: number;
    total: number;
  };
}
