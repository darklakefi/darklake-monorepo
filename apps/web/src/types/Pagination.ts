export type PaginatedResponse<T> = {
  limit: number;
  offset: number;
  total: number;
  result: T;
};

export enum SortDirection {
  ASC = "asc",
  DESC = "desc",
}
