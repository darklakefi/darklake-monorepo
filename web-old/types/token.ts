export interface Token {
  symbol: string;
  name: string;
  image: string;
  address: string;
  tokenProgram: string;
  decimals: number;
}
