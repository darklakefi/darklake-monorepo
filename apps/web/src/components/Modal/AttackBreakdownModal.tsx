import Modal, { ModalProps } from "@/components/Modal";
import { MevAttack, MevAttackSwapType } from "@/types/Mev";
import { format } from "date-fns";
import { formatMoney, formatPercentage } from "@/utils/number";
import { cn, truncate } from "@/utils/common";

interface AttackBreakdownModalProps extends ModalProps {
  mevAttack?: MevAttack;
}

export default function AttackBreakdownModal(props: AttackBreakdownModalProps) {
  const { mevAttack } = props;

  if (!mevAttack?.transactions || !mevAttack?.solAmount) {
    return (
      <Modal title="Attack breakdown" {...props}>
        <div className="lg:w-[960px] p-5 text-brand-30">
          <p>Unable to load attack details. Please try again later.</p>
        </div>
      </Modal>
    );
  }

  const isVictimBuy = mevAttack.swapType === MevAttackSwapType.BUY;
  const solAmountLostFormatted = formatMoney(mevAttack.solAmount.lost);
  const solAmountSentFormatted = formatMoney(mevAttack.solAmount.sent);

  return (
    <Modal title="Attack breakdown" {...props}>
      <div className="lg:w-[960px]">
        <div className="uppercase text-lg leading-6 tracking-normal text-brand-30 mb-6">
          <p>Token: {mevAttack.tokenName}</p>
          <p>Date: {format(mevAttack.timestamp, "yyyy-MM-dd hh:mm OOO")}</p>
        </div>
        <div className="flex flex-row justify-between items-center p-5 bg-brand-60 mb-6">
          {[
            { title: "Total lost", value: `${solAmountLostFormatted} SOL` },
            {
              title: "Extracted",
              value: `${formatPercentage((mevAttack.solAmount.lost / mevAttack.solAmount.sent) * 100)}%`,
            },
            { title: "TX size", value: `${solAmountSentFormatted} SOL` },
          ].map((row) => (
            <div key={row.title} className="uppercase">
              <p className="text-3xl leading-9 text-brand-20">{row.value}</p>
              <p className="text-lg leading-6 tracking-normal text-brand-30">{row.title}</p>
            </div>
          ))}
        </div>
        <div className="relative">
          <span className="w-px h-full absolute left-[11.5px] max-md:hidden top-0 bg-brand-50" />
          {[
            {
              title: "Frontrun Attack",
              iconClassName: "hn-exclamation-triangle-solid",
              details:
                `Attacker ${isVictimBuy ? "buys" : "sells"} ${mevAttack.tokenName} ` +
                `tokens to drive ${isVictimBuy ? "up" : "down"} price`,
              transaction: mevAttack.transactions.frontRun,
            },
            {
              title: "Victim Transaction",
              iconClassName: "hn-refresh-solid",
              details:
                `Swap: ${isVictimBuy ? `${solAmountSentFormatted} SOL` : mevAttack.tokenName} ` +
                `→ ${isVictimBuy ? mevAttack.tokenName : `${solAmountSentFormatted} SOL`} ` +
                `– executed at ${isVictimBuy ? "inflated" : "dropped"} price`,
              transaction: mevAttack.transactions.victim,
            },
            {
              title: "Backrun Attack",
              iconClassName: "hn-exclamation-triangle-solid",
              details:
                `Attacker ${isVictimBuy ? "sells" : "buys"} ${mevAttack.tokenName} ` +
                `tokens at ${isVictimBuy ? "peak" : "floor"} price ` +
                `– value extracted: ${solAmountLostFormatted} SOL`,
              transaction: mevAttack.transactions.frontRun,
            },
          ].map((row) => (
            <div
              key={`${row.title}-${row.transaction.address}${row.transaction.address}-${row.transaction.signature}`}
              className={cn(
                "md:flex flex-row items-center w-full gap-x-2 md:mb-6 relative",
                "uppercase text-lg leading-5",
                "group",
              )}
            >
              <span className="hidden md:group-first-of-type:block absolute top-0 left-0 bg-brand-70 h-1/2 w-full" />
              <span className="hidden md:group-last-of-type:block absolute bottom-0 left-0 bg-brand-70 h-1/2 w-full" />
              <div
                className={cn(
                  "text-brand-20 md:w-[264px]",
                  "flex flex-row items-center justify-center py-5 bg-brand-70 relative",
                )}
              >
                <i className={cn("text-[22px] hn mr-4", row.iconClassName)} />
                <p>{row.title}</p>
                <span className="max-md:hidden ml-1 h-px w-20 bg-brand-50 flex-1" />
              </div>
              <div className=" text-brand-30 border border-brand-60 p-5 flex-1 relative">
                <p className="text-brand-20">{row.details}</p>
                <p>{truncate(row.transaction.signature)}</p>
                <a
                  href={`${process.env.NEXT_PUBLIC_EXPLORER_URL}/tx/${row.transaction.signature}`}
                  className="underline"
                  title="View Transaction"
                  target="_blank"
                >
                  View TX
                </a>
              </div>
              <span className="block group-last-of-type:hidden md:hidden w-px mt-4 h-8 mx-auto bg-brand-50" />
            </div>
          ))}
        </div>
      </div>
    </Modal>
  );
}
