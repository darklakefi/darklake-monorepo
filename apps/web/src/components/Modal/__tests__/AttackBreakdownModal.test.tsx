import { act, render, RenderResult } from "@testing-library/react";
import AttackBreakdownModal from "@/components/Modal/AttackBreakdownModal";
import { MevAttack, MevAttackSwapType, MevTransaction } from "@/types/Mev";

const mockMevAttackTransaction: MevTransaction = {
  address: "someAddress",
  signature: "someSignature",
};

const mockMevAttack: MevAttack = {
  swapType: MevAttackSwapType.BUY,
  timestamp: +new Date("2024-03-26"),
  tokenName: "TRUMP",
  solAmount: {
    sent: 100,
    lost: 30,
  },
  transactions: {
    frontRun: mockMevAttackTransaction,
    victim: mockMevAttackTransaction,
    backRun: mockMevAttackTransaction,
  },
};

describe("<AttackBreakdownModal />", () => {
  it("does not display content when closed", async () => {
    let baseElement: RenderResult["baseElement"] | undefined;

    await act(async () => {
      ({ baseElement } = render(<AttackBreakdownModal mevAttack={mockMevAttack} isOpen={false} />));
    });

    expect(baseElement).toMatchSnapshot();
  });

  it("displays content successfully when open", async () => {
    let baseElement: RenderResult["baseElement"] | undefined;

    await act(async () => {
      ({ baseElement } = render(<AttackBreakdownModal mevAttack={mockMevAttack} isOpen />));
    });

    expect(baseElement).toMatchSnapshot();
  });
});
