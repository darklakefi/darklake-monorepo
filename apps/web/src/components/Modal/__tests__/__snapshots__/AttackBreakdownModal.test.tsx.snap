// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AttackBreakdownModal /> displays content successfully when open 1`] = `
<body>
  <div />
  <div
    id="headlessui-portal-root"
  >
    <div
      data-headlessui-portal=""
    >
      <button
        aria-hidden="true"
        data-headlessui-focus-guard="true"
        style="position: fixed; top: 1px; left: 1px; width: 1px; height: 0px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px;"
        type="button"
      />
      <div>
        <div
          aria-modal="true"
          class="relative"
          data-headlessui-state="open"
          data-open=""
          id="headlessui-dialog-:r0:"
          role="dialog"
          tabindex="-1"
        >
          <div
            aria-hidden="true"
            class="fixed inset-0 bg-modal-backdrop"
            data-headlessui-state="open"
            data-open=""
          />
          <div
            class="flex flex-col fixed inset-0 w-screen items-center justify-center"
          >
            <div
              data-headlessui-state="open"
              data-open=""
              id="headlessui-dialog-panel-:r7:"
            >
              <div
                class="py-4 pl-4 pr-7"
              >
                <div
                  class="max-h-[80vh] lg:max-h-none overflow-auto p-4 md:p-10 bg-brand-70 app-box-shadow relative"
                >
                  <div
                    class="flex flex-row items-center mb-4 border-b border-b-brand-50 pb-3 justify-between"
                  >
                    <p
                      class="font-primary text-3xl leading-3xl text-brand-20"
                    >
                      Attack breakdown
                    </p>
                    <button
                      class="hover-with-active flex items-center"
                      type="button"
                    >
                      <i
                        class="hn hn-times text-brand-30 text-xl leading-5"
                      />
                    </button>
                  </div>
                  <div
                    class="font-secondary text-lg leading-6 text-brand-30"
                  >
                    <div
                      class="lg:w-[960px]"
                    >
                      <div
                        class="uppercase text-lg leading-6 text-brand-30 mb-6"
                      >
                        <p>
                          Token: 
                          TRUMP
                        </p>
                        <p>
                          Date: 
                          2024-03-26 12:00 GMT+0
                        </p>
                      </div>
                      <div
                        class="flex flex-row justify-between items-center p-5 bg-brand-60 mb-6"
                      >
                        <div
                          class="uppercase"
                        >
                          <p
                            class="text-3xl leading-9 text-brand-20"
                          >
                            30 SOL
                          </p>
                          <p
                            class="text-lg leading-6 text-brand-30"
                          >
                            Total lost
                          </p>
                        </div>
                        <div
                          class="uppercase"
                        >
                          <p
                            class="text-3xl leading-9 text-brand-20"
                          >
                            30%
                          </p>
                          <p
                            class="text-lg leading-6 text-brand-30"
                          >
                            Extracted
                          </p>
                        </div>
                        <div
                          class="uppercase"
                        >
                          <p
                            class="text-3xl leading-9 text-brand-20"
                          >
                            100 SOL
                          </p>
                          <p
                            class="text-lg leading-6 text-brand-30"
                          >
                            TX size
                          </p>
                        </div>
                      </div>
                      <div
                        class="relative"
                      >
                        <span
                          class="w-px h-full absolute left-[11.5px] max-md:hidden top-0 bg-brand-50"
                        />
                        <div
                          class="md:flex flex-row items-center w-full gap-x-[7px] md:mb-6 relative uppercase text-lg leading-5 group"
                        >
                          <span
                            class="hidden md:group-first-of-type:block absolute top-0 left-0 bg-brand-70 h-1/2 w-full"
                          />
                          <span
                            class="hidden md:group-last-of-type:block absolute bottom-0 left-0 bg-brand-70 h-1/2 w-full"
                          />
                          <div
                            class="text-brand-20 md:w-[264px] flex flex-row items-center justify-center py-[18px] bg-brand-70 relative"
                          >
                            <i
                              class="text-[22px] hn mr-4 hn-exclamation-triangle-solid"
                            />
                            <p>
                              Frontrun Attack
                            </p>
                            <span
                              class="max-md:hidden ml-[5px] h-px w-[74px] bg-brand-50 flex-1"
                            />
                          </div>
                          <div
                            class=" text-brand-30 border border-brand-60 p-5 flex-1 relative"
                          >
                            <p
                              class="text-brand-20"
                            >
                              Attacker buys TRUMP tokens to drive up price
                            </p>
                            <p>
                              someSignature
                            </p>
                            <a
                              class="underline"
                              href="#"
                              target="_blank"
                              title="View Transaction"
                            >
                              View TX
                            </a>
                          </div>
                          <span
                            class="block group-last-of-type:hidden md:hidden w-px mt-4 h-8 mx-auto bg-brand-50"
                          />
                        </div>
                        <div
                          class="md:flex flex-row items-center w-full gap-x-[7px] md:mb-6 relative uppercase text-lg leading-5 group"
                        >
                          <span
                            class="hidden md:group-first-of-type:block absolute top-0 left-0 bg-brand-70 h-1/2 w-full"
                          />
                          <span
                            class="hidden md:group-last-of-type:block absolute bottom-0 left-0 bg-brand-70 h-1/2 w-full"
                          />
                          <div
                            class="text-brand-20 md:w-[264px] flex flex-row items-center justify-center py-[18px] bg-brand-70 relative"
                          >
                            <i
                              class="text-[22px] hn mr-4 hn-refresh-solid"
                            />
                            <p>
                              Victim Transaction
                            </p>
                            <span
                              class="max-md:hidden ml-[5px] h-px w-[74px] bg-brand-50 flex-1"
                            />
                          </div>
                          <div
                            class=" text-brand-30 border border-brand-60 p-5 flex-1 relative"
                          >
                            <p
                              class="text-brand-20"
                            >
                              Swap: 100 SOL → TRUMP – executed at inflated price
                            </p>
                            <p>
                              someSignature
                            </p>
                            <a
                              class="underline"
                              href="#"
                              target="_blank"
                              title="View Transaction"
                            >
                              View TX
                            </a>
                          </div>
                          <span
                            class="block group-last-of-type:hidden md:hidden w-px mt-4 h-8 mx-auto bg-brand-50"
                          />
                        </div>
                        <div
                          class="md:flex flex-row items-center w-full gap-x-[7px] md:mb-6 relative uppercase text-lg leading-5 group"
                        >
                          <span
                            class="hidden md:group-first-of-type:block absolute top-0 left-0 bg-brand-70 h-1/2 w-full"
                          />
                          <span
                            class="hidden md:group-last-of-type:block absolute bottom-0 left-0 bg-brand-70 h-1/2 w-full"
                          />
                          <div
                            class="text-brand-20 md:w-[264px] flex flex-row items-center justify-center py-[18px] bg-brand-70 relative"
                          >
                            <i
                              class="text-[22px] hn mr-4 hn-exclamation-triangle-solid"
                            />
                            <p>
                              Backrun Attack
                            </p>
                            <span
                              class="max-md:hidden ml-[5px] h-px w-[74px] bg-brand-50 flex-1"
                            />
                          </div>
                          <div
                            class=" text-brand-30 border border-brand-60 p-5 flex-1 relative"
                          >
                            <p
                              class="text-brand-20"
                            >
                              Attacker sells TRUMP tokens at peak price – value extracted: 30 SOL
                            </p>
                            <p>
                              someSignature
                            </p>
                            <a
                              class="underline"
                              href="#"
                              target="_blank"
                              title="View Transaction"
                            >
                              View TX
                            </a>
                          </div>
                          <span
                            class="block group-last-of-type:hidden md:hidden w-px mt-4 h-8 mx-auto bg-brand-50"
                          />
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <button
        aria-hidden="true"
        data-headlessui-focus-guard="true"
        style="position: fixed; top: 1px; left: 1px; width: 1px; height: 0px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px;"
        type="button"
      />
    </div>
  </div>
</body>
`;

exports[`<AttackBreakdownModal /> does not display content when closed 1`] = `
<body>
  <div />
</body>
`;
